File: workflows/simulation/run_simulation.jl
#=
NEA/EHI Stochastic Epidemiological Simulation — Julia Core
Called from run_simulation.R via system() or JuliaCall

State space: 4,000+ variables
- 66 age groups × 62 disease components = 4,092 human states
- 11 mosquito compartment states
- Daily timestep, 36,500-day simulation (100 years)

This is a placeholder to be replaced with Chia-Chen CHANG's actual model.
=#

using DifferentialEquations
using Distributions
using Random
using JSON
using CSV
using DataFrames

# ============================================================
# Model Parameters
# ============================================================
struct DengueParams
    n_age_groups::Int          # 66
    n_disease_components::Int  # 62
    n_mosquito_components::Int # 11
    beta::Float64              # Transmission rate
    mu_h::Float64              # Human mortality rate
    mu_v::Float64              # Mosquito mortality rate
    gamma::Float64             # Recovery rate
    sigma::Float64             # Incubation rate
    alpha::Vector{Float64}     # Age-specific contact rates
end

function default_params()
    n_age = 66
    DengueParams(
        n_age,
        62,
        11,
        0.5,    # beta
        1/25550, # mu_h (~70 years)
        1/14,   # mu_v (~14 days)
        1/7,    # gamma (~7 days recovery)
        1/5.5,  # sigma (~5.5 days incubation)
        ones(n_age) .* 0.1  # uniform contact rates (placeholder)
    )
end

# ============================================================
# State Vector Layout
# ============================================================
# Human states: [age_group, disease_component]
# Mosquito states: [compartment]
# Total: 66 * 62 + 11 = 4,103

function state_size(p::DengueParams)
    p.n_age_groups * p.n_disease_components + p.n_mosquito_components
end

# ============================================================
# Stochastic Model (SDE)
# ============================================================
function dengue_drift!(du, u, p::DengueParams, t)
    n_human = p.n_age_groups * p.n_disease_components
    n_total = state_size(p)

    # Human compartment dynamics (placeholder)
    for i in 1:n_human
        age = ((i - 1) % p.n_age_groups) + 1
        du[i] = -p.mu_h * u[i] + p.sigma * max(0, u[max(1, i-1)]) - p.gamma * u[i]
    end

    # Mosquito dynamics (placeholder)
    for j in (n_human+1):n_total
        du[j] = p.beta * u[j] - p.mu_v * u[j]
    end
end

function dengue_diffusion!(du, u, p::DengueParams, t)
    n_total = state_size(p)
    for i in 1:n_total
        du[i] = sqrt(max(0.0, abs(u[i]))) * 0.01  # Small stochastic perturbation
    end
end

# ============================================================
# Single Simulation Run
# ============================================================
function run_single(params::DengueParams, n_days::Int, seed::Int)
    Random.seed!(seed)
    n = state_size(params)

    # Initial conditions
    u0 = zeros(Float64, n)
    n_human = params.n_age_groups * params.n_disease_components

    # Initialize susceptible population across age groups
    for age in 1:params.n_age_groups
        u0[age] = 1000.0  # Susceptible pool
    end

    # Initialize mosquito compartments
    for j in (n_human+1):n
        u0[j] = 500.0  # Mosquito population
    end

    tspan = (0.0, Float64(n_days))
    prob = SDEProblem(dengue_drift!, dengue_diffusion!, u0, tspan, params)

    sol = solve(prob, EM(), dt=1.0, saveat=1.0)

    return sol
end

# ============================================================
# Parameter Estimation (Metropolis-Hastings MCMC)
# ============================================================
function run_mcmc_chain(
    params::DengueParams,
    observed_data::Vector{Float64},
    chain_index::Int,
    n_iterations::Int,
    seed::Int
)
    Random.seed!(seed + chain_index)

    # Parameter vector to estimate: [beta, gamma, sigma, mu_v]
    current = [params.beta, params.gamma, params.sigma, params.mu_v]
    n_params = length(current)

    # Proposal standard deviations
    proposal_sd = [0.05, 0.02, 0.02, 0.005]

    # Storage
    chain = zeros(Float64, n_iterations, n_params)
    log_likelihoods = zeros(Float64, n_iterations)
    accept_count = 0

    current_ll = -Inf

    for iter in 1:n_iterations
        # Propose
        proposed = current .+ randn(n_params) .* proposal_sd

        # Ensure positivity
        if any(proposed .<= 0)
            chain[iter, :] = current
            log_likelihoods[iter] = current_ll
            continue
        end

        # Evaluate (placeholder — simplified likelihood)
        proposed_ll = -sum((observed_data .- proposed[1] .* 100) .^ 2) / (2 * 100^2)

        # Accept/reject
        log_alpha = proposed_ll - current_ll
        if log(rand()) < log_alpha
            current = proposed
            current_ll = proposed_ll
            accept_count += 1
        end

        chain[iter, :] = current
        log_likelihoods[iter] = current_ll

        if iter % 500 == 0
            acceptance_rate = accept_count / iter
            println("Chain $chain_index | Iter $iter/$n_iterations | Accept rate: $(round(acceptance_rate, digits=3))")
        end
    end

    acceptance_rate = accept_count / n_iterations
    println("Chain $chain_index complete. Final acceptance rate: $(round(acceptance_rate, digits=3))")

    return Dict(
        "chain" => chain_index,
        "samples" => chain,
        "log_likelihoods" => log_likelihoods,
        "acceptance_rate" => acceptance_rate,
        "n_iterations" => n_iterations
    )
end

# ============================================================
# Main Entrypoint
# ============================================================
function main()
    run_mode = get(ENV, "RUN_MODE", "single")
    chain_index = parse(Int, get(ENV, "AWS_BATCH_JOB_ARRAY_INDEX", "0"))
    n_iterations = parse(Int, get(ENV, "ITERATIONS", "5000"))
    seed = 42

    params = default_params()
    println("State vector size: $(state_size(params))")

    if run_mode == "single"
        println("=== Running single simulation (36,500 days) ===")
        sol = run_single(params, 36500, seed)
        println("Simulation complete. Timepoints: $(length(sol.t))")

        # Save summary to JSON
        result = Dict(
            "mode" => "single",
            "n_days" => 36500,
            "n_states" => state_size(params),
            "n_timepoints" => length(sol.t),
            "final_state_summary" => Dict(
                "mean" => mean(sol.u[end]),
                "max" => maximum(sol.u[end]),
                "min" => minimum(sol.u[end])
            )
        )
        open("/tmp/julia_result.json", "w") do f
            JSON.print(f, result)
        end

    elseif run_mode == "param-estimation"
        println("=== Running MCMC parameter estimation ===")
        println("Chain: $chain_index | Iterations: $n_iterations")

        # Placeholder observed data
        observed = randn(100) .* 30 .+ 50

        result = run_mcmc_chain(params, observed, chain_index, n_iterations, seed)

        # Save chain to JSON
        open("/tmp/julia_result.json", "w") do f
            JSON.print(f, Dict(
                "mode" => "param-estimation",
                "chain" => chain_index,
                "iterations" => n_iterations,
                "acceptance_rate" => result["acceptance_rate"],
                "posterior_mean" => vec(mean(result["samples"], dims=1))
            ))
        end
    else
        error("Unknown RUN_MODE: $run_mode")
    end

    println("=== Julia simulation runner complete ===")
end

# Run
main()

File: workflows/simulation/run_simulation.R
#!/usr/bin/env Rscript
# NEA/EHI Stochastic Simulation Runner
# Entrypoint for AWS Batch jobs
# Supports: single-run mode and parameter estimation mode

library(DBI)
library(RPostgres)
library(jsonlite)
library(dplyr)

# Environment variables set by AWS Batch job definition
s3_bucket    <- Sys.getenv("S3_BUCKET")
run_mode     <- Sys.getenv("RUN_MODE", "single")     # "single" or "param-estimation"
num_chains   <- as.integer(Sys.getenv("NUM_CHAINS", "4"))
iterations   <- as.integer(Sys.getenv("ITERATIONS", "5000"))
chain_index  <- as.integer(Sys.getenv("AWS_BATCH_JOB_ARRAY_INDEX", "0"))

cat(sprintf("=== NEA/EHI Simulation Runner ===\n"))
cat(sprintf("Mode: %s | Chain: %d | Iterations: %d\n", run_mode, chain_index, iterations))

# --- Database Connection ---
get_db_connection <- function() {
  secret_json <- system("aws secretsmanager get-secret-value --secret-id nea-ehi-poc/rds-master --query SecretString --output text", intern = TRUE)
  creds <- fromJSON(secret_json)

  dbConnect(
    RPostgres::Postgres(),
    host     = creds$host,
    port     = as.integer(creds$port),
    dbname   = "nea_ehi",
    user     = creds$username,
    password = creds$password,
    sslmode  = "require"
  )
}

# --- Load Parameters from RDS ---
load_parameters <- function(con) {
  # Load simulation parameters from database
  params <- dbGetQuery(con, "
    SELECT * FROM simulation_params
    WHERE active = TRUE
    ORDER BY created_at DESC
    LIMIT 1
  ")

  if (nrow(params) == 0) {
    stop("No active simulation parameters found in database")
  }

  fromJSON(params$params_json[1])
}

# --- Simulation Core ---
# 4,000 state variables: 66 age groups x 62 disease components + 11 mosquito components
run_single_simulation <- function(params, seed = NULL) {
  if (!is.null(seed)) set.seed(seed)

  start_time <- Sys.time()
  cat(sprintf("Starting simulation at %s\n", start_time))

  # Placeholder for actual NEA simulation code
  # This will be replaced with Chia-Chen's R + Julia simulation logic
  n_days <- 36500  # 100 years, daily timestep
  n_age_groups <- 66
  n_disease_components <- 62
  n_mosquito_components <- 11
  n_states <- n_age_groups * n_disease_components + n_mosquito_components  # ~4,103

  # Initialize state vector
  state <- rep(0, n_states)

  # Simulation loop (placeholder — actual ODE/stochastic logic from NEA team)
  results <- list()
  for (day in seq_len(n_days)) {
    # Daily state update (placeholder)
    state <- state + rnorm(n_states, mean = 0, sd = 0.01)

    # Save daily snapshot
    if (day %% 365 == 0) {
      results[[length(results) + 1]] <- list(
        day = day,
        summary = summary(state)
      )
    }
  }

  elapsed <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))
  cat(sprintf("Simulation completed in %.1f seconds\n", elapsed))

  list(
    results = results,
    runtime_sec = elapsed,
    n_days = n_days,
    n_states = n_states
  )
}

# --- Parameter Estimation (MCMC) ---
run_parameter_estimation <- function(params, chain_idx, n_iter) {
  start_time <- Sys.time()
  cat(sprintf("Starting parameter estimation: chain %d, %d iterations\n", chain_idx, n_iter))

  set.seed(42 + chain_idx)

  # Placeholder for MCMC parameter estimation
  # Will be replaced with Chia-Chen's actual estimation code
  posterior_samples <- matrix(rnorm(n_iter * 10), nrow = n_iter, ncol = 10)
  colnames(posterior_samples) <- paste0("param_", 1:10)

  elapsed <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))
  cat(sprintf("Parameter estimation chain %d completed in %.1f seconds\n", chain_idx, elapsed))

  list(
    chain = chain_idx,
    posterior = posterior_samples,
    runtime_sec = elapsed,
    iterations = n_iter
  )
}

# --- Upload Results to S3 ---
upload_to_s3 <- function(result, run_id) {
  output_path <- sprintf("/tmp/%s.json", run_id)
  writeLines(toJSON(result, auto_unbox = TRUE), output_path)

  s3_key <- sprintf("results/simulation/%s.json", run_id)
  system(sprintf("aws s3 cp %s s3://%s/%s --sse aws:kms", output_path, s3_bucket, s3_key))
  cat(sprintf("Results uploaded to s3://%s/%s\n", s3_bucket, s3_key))

  s3_key
}

# --- Record Run in RDS ---
record_run <- function(con, run_id, model_type, params_json, status, result_path, runtime_sec) {
  dbExecute(con, "
    INSERT INTO simulation_runs (run_id, model_type, params, status, result_path, runtime_sec)
    VALUES ($1, $2, $3::jsonb, $4, $5, $6)
  ", list(run_id, model_type, params_json, status, result_path, runtime_sec))
}

# --- Main ---
main <- function() {
  con <- get_db_connection()
  on.exit(dbDisconnect(con))

  params <- load_parameters(con)
  run_id <- sprintf("sim_%s_%s_%d", run_mode, format(Sys.time(), "%Y%m%d_%H%M%S"), chain_index)

  tryCatch({
    if (run_mode == "single") {
      result <- run_single_simulation(params)
    } else if (run_mode == "param-estimation") {
      result <- run_parameter_estimation(params, chain_index, iterations)
    } else {
      stop(sprintf("Unknown run mode: %s", run_mode))
    }

    s3_path <- upload_to_s3(result, run_id)
    record_run(con, run_id, run_mode, toJSON(params, auto_unbox = TRUE), "completed", s3_path, result$runtime_sec)
    cat("=== Simulation completed successfully ===\n")
  }, error = function(e) {
    cat(sprintf("ERROR: %s\n", e$message))
    record_run(con, run_id, run_mode, toJSON(params, auto_unbox = TRUE), "failed", NA, NA)
    quit(status = 1)
  })
}

main()
